nncf/experimental/common/pruning/operations.py

```diff
@@ -0,0 +1,637 @@
+
+        node.data['output_mask'] = output_mask
+
+
+class ReshapeMode(Enum):
+    """
+    Defines the mode of reshape.
+    Here's examples of reshaping for each mode:
+        Extend: [N,C,H,W] ----(reshaped to)---> [N, C1, C2, H, W1, W2], when C=C1*C2 and W=W1*W2
+        Srink is opposite to Extend: [N, C1, C2, H, W1, W2] ----(reshaped to)---> [N,C,H,W]
+        Default - all other cases.
+    """
+    SHRINK = auto()
+    EXTEND = auto()
+    DEFAULT = auto()
+
+
+class ReshapePruningOp(BasePruningOp):
+    DIMENSION_MAP = Dict[int, int]
+
+    @classmethod
+    def accept_pruned_input(cls, node: NNCFNode) -> bool:
+        if node.layer_attributes is None:
+            return False
+        return True
+
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        node.data['output_mask'] = cls._get_propagated_mask(node, graph)
+
+    @staticmethod
+    def _check_dim_splitted(dim_from: int,
+                            dims_to: List[int],
+                            dims_to_start_idx: int) -> Tuple[bool, int]:
+        idx = dims_to_start_idx
+        accum = dims_to[idx]
+        while accum < dim_from:
+            idx += 1
+            accum *= dims_to[idx]
+        if accum > dim_from:
+            return (False, idx)
+        return (True, idx)
+
+    @classmethod
+    def _map_dims_(cls,
+                   dims_from: List[int],
+                   dims_to: List[int],
+                   from_idx: int,
+                   to_idx: int,
+                   from_map: DIMENSION_MAP,
+                   to_map: DIMENSION_MAP) -> Tuple[bool, int]:
+        res, to_idx_next = cls._check_dim_splitted(dims_from[from_idx], dims_to, to_idx)
+        if not res:
+            return (res, to_idx_next)
+        from_map[from_idx] = list(range(to_idx, to_idx_next + 1))
+        for idx in range(to_idx, to_idx_next + 1):
+            to_map[idx] = from_idx
+        return (res, to_idx_next)
+
+    @classmethod
+    def _parse_reshape(cls, node: NNCFNode) -> Tuple[DIMENSION_MAP, DIMENSION_MAP, ReshapeMode]:
+        input_shape = node.layer_attributes.input_shape
+        output_shape = node.layer_attributes.output_shape
+
+        inp_idx = 0
+        out_idx = 0
+        inp_map = {}
+        out_map = {}
+
+        mode = ReshapeMode.DEFAULT
+
+        while (inp_idx < len(input_shape) and out_idx < len(output_shape)):
+            if input_shape[inp_idx] == output_shape[out_idx]:
+                inp_map[inp_idx] = out_idx
+                out_map[out_idx] = inp_idx
+            elif input_shape[inp_idx] > output_shape[out_idx]:
+                res, out_idx = cls._map_dims_(input_shape, output_shape,
+                                              inp_idx, out_idx, inp_map, out_map)
+                assert res and mode != ReshapeMode.SHRINK, "Invalid reshape parsing"
```

Comment:
This shouldn't be invalid reshape parsing, this expression means that algorithm doesn't support mask propagation through such reshape and all masks should  be invalidated.

Reply:
Done